ha.go

```diff
@@ -0,0 +1,85 @@
+package rockset
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/rockset/rockset-go-client/openapi"
+	"github.com/rockset/rockset-go-client/option"
+)
+
+type Querier interface {
+	Query(context.Context, string, ...option.QueryOption) (openapi.QueryResponse, error)
+}
+
+type HA struct {
+	clients []Querier
+}
+
+func NewHA(client ...Querier) *HA {
+	return &HA{client}
+}
+
+func (ha *HA) Query(ctx context.Context, query string, options ...option.QueryOption) (openapi.QueryResponse, []error) {
+	log := zerolog.Ctx(ctx)
+	// create a sub-context, so we can cancel all HTTP requests after getting the first answer
+	subCtx, cancel := context.WithCancel(ctx)
+
+	// make sure to cancel any pending query result as we don't need them when we return
+	defer cancel()
+
+	var wg sync.WaitGroup
+	resultCh := make(chan openapi.QueryResponse, len(ha.clients))
+	errorCh := make(chan error, len(ha.clients))
+
+	wg.Add(len(ha.clients))
+	for idx, c := range ha.clients {
+		go func(i int, cl Querier) {
+			res, err := cl.Query(ctx, query, options...)
+			if err != nil {
+				log.Error().Err(err).Int("idx", i).Msg("failed to query")
+				errorCh <- err
+			} else {
+				log.Trace().Int("idx", i).Msg("got query response")
+				resultCh <- res
+			}
+			wg.Done()
+		}(idx, c) // avoid using the loop variables as they mutate each iteration
+	}
+
+	// wait for all queries to be done (or cancelled) and then close the channels
```

Comment:
don't you want the first query that returns successfully? otherwise you're always paying the cost of higher latency to another region

Reply:
The comment is misleading. It returns the first result, as soon as it gets it, but there is some cleanup happening so it doesn't leak any go routines or channels.

Once it gets a result, line 67 is executed, which triggers the defer on line 31 to run. That'll cause all remaining go calls to abort which in turn trigger this go routine to close all channels.